File: 0399-evaluate-division/0399-evaluate-division.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        ans = []
        graph = defaultdict(dict)
        for i in range(len(equations)):
            a, b = equations[i]
            value = values[i]
            graph[a][b] = value
            graph[b][a] = 1.0 / value
        for query in queries:
            a, b = query
            visited = set()
            result = self.dfs(a, b, graph, visited)
            if result != 0.0:
                ans.append(result)
            else:
                ans.append(-1.0)
        return ans
    def dfs(self, a, b, graph, visited):
        if a not in graph or b not in graph:
            return 0.0
        if b in graph[a]:
            return graph[a][b]
        visited.add(a)
        for node, value in graph[a].items():
            if node not in visited:
                result = self.dfs(node, b, graph, visited)
                if result != 0.0:
                    graph[a][b] = value * result
                    return graph[a][b]
        return 0.0

File: 0399-evaluate-division/0399-evaluate-division.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        List<Double> ans = new ArrayList<>();
        Map<String, Map<String, Double>> graph = new HashMap<>();
        for (int i = 0; i < equations.size(); i++) {
            String a = equations.get(i).get(0);
            String b = equations.get(i).get(1);
            double value = values[i];

            graph.putIfAbsent(a, new HashMap<>());
            graph.putIfAbsent(b, new HashMap<>());

            graph.get(a).put(b, value);
            graph.get(b).put(a, 1.0 / value);
        }
        for (int i = 0; i < queries.size(); i++) {
            String a = queries.get(i).get(0);
            String b = queries.get(i).get(1);
            Set<String> visited = new HashSet<>();
            double val = dfs(a, b, graph, visited);
            if (val != 0.0) ans.add(val);
            else ans.add(-1.0);
        }
        double[] result = new double[ans.size()];
        for (int i = 0; i < ans.size(); i++) {
            result[i] = ans.get(i);
        }
        return result;
    }
    double dfs(String a, String b, Map<String, Map<String, Double>> graph, Set<String> visited){
        if (!graph.containsKey(a) || !graph.containsKey(b)) return 0.0;
        if (graph.get(a).containsKey(b)) return graph.get(a).get(b);
        visited.add(a);
        for (Map.Entry<String, Double> entry : graph.get(a).entrySet()) {
            String nextNode = entry.getKey();
            double value = entry.getValue();
            if (!visited.contains(nextNode)) {
                double result = dfs(nextNode, b, graph, visited);
                if (result != 0.0) {
                    graph.get(a).put(b, value * result);
                    return graph.get(a).get(b);
                }
            }
        }
        return 0.0;
    }
}